File: controllers/visiteurController.js
/**
 * Construit un regex flexible pour faire correspondre les numéros de téléphone
 * peu importe les espaces, tirets ou indicatifs (+221, 00221, etc.)
 */
const construireRegexTelephone = (queryTel) => {
  if (!queryTel) return null;
  const digitsOnly = String(queryTel).replace(/\D/g, '');
  if (digitsOnly.length < 6) return null;
  const lastDigits = digitsOnly.slice(-8);
  const regexPattern = lastDigits.split('').join('[\\s.-]*');
  return new RegExp(regexPattern, 'i');
};

/**
 * Recherche un visiteur existant par son NIN (Numéro d'Identification National) ou numéro de pièce
 */
const rechercherParNIN = async (req, res) => {
  try {
    const queryNin = req.query.nin || req.query.q || req.query.telephone || req.query.phone;
    if (!queryNin) {
      return res.status(400).json({ success: false, message: 'Le NIN est requis.' });
    }

    const rawQuery = String(queryNin).trim();
    const digitsOnly = rawQuery.replace(/\D/g, '');

    const filtre = {
      $or: [
        { nin: rawQuery },
        ...(digitsOnly.length >= 5 ? [{ nin: { $regex: digitsOnly, $options: 'i' } }] : []),
        { numeroPiece: rawQuery }
      ]
    };

    const visiteur = await Visiteur.findOne(filtre);

    if (!visiteur) {
      return res.status(404).json({
        success: false,
        message: 'Aucun visiteur trouvé avec ce NIN.'
      });
    }

    return res.json({
      success: true,
      message: 'Visiteur existant trouvé dans la base de données.',
      visiteur
    });
  } catch (err) {
    return res.status(500).json({ success: false, message: err.message });
  }
};

const creerVisiteur = async (req, res) => {
  try {
    const {
      nom, prenom, dateNaissance, lieuNaissance, sexe, taille,
      numeroPiece, typePiece, dateDelivrance, dateExpiration,
      centreEnregistrement, adresseDomicile,
      nin, telephone
    } = req.body;

    let cleanNumPiece = numeroPiece ? String(numeroPiece).trim() : null;
    let cleanNin = nin ? String(nin).trim() : null;

    if (!cleanNumPiece || cleanNumPiece.toLowerCase() === 'undefined' || cleanNumPiece.toLowerCase() === 'null') {
      if (cleanNin) {
        cleanNumPiece = `NIN-${cleanNin.replace(/\D/g, '')}`;
      } else if (telephone) {
        cleanNumPiece = `TEL-${String(telephone).replace(/\D/g, '')}`;
      } else {
        cleanNumPiece = `VIS-${Date.now()}`;
      }
    }

    let existeDeja = null;

    // 1. Recherche prioritaire par NIN
    if (cleanNin) {
      const ninDigits = cleanNin.replace(/\D/g, '');
      existeDeja = await Visiteur.findOne({
        $or: [
          { nin: cleanNin },
          ...(ninDigits.length >= 6 ? [{ nin: { $regex: ninDigits, $options: 'i' } }] : [])
        ]
      });
    }

    // 2. Fallback recherche par Numéro de pièce
    if (!existeDeja && cleanNumPiece) {
      existeDeja = await Visiteur.findOne({ numeroPiece: cleanNumPiece });
    }

    if (existeDeja) {
      let modified = false;
      if (cleanNin && !existeDeja.nin) {
        existeDeja.nin = cleanNin;
        modified = true;
      }
      if (telephone && !existeDeja.telephone) {
        existeDeja.telephone = telephone;
        modified = true;
      }
      if (modified) await existeDeja.save();

      return res.status(200).json({
        success: true,
        message: 'Visiteur déjà enregistré.',
        visiteur: existeDeja,
        estNouveau: false,
        id: existeDeja._id
      });
    }

    const visiteur = await Visiteur.create({
      nom: nom || 'Visiteur',
      prenom: prenom || 'Anonyme',
      dateNaissance,
      lieuNaissance,
      sexe,
      taille,
      numeroPiece: cleanNumPiece,
      typePiece: typePiece || 'CNI',
      dateDelivrance,
      dateExpiration,
      centreEnregistrement,
      adresseDomicile,
      nin: cleanNin,
      telephone
    });

    res.status(201).json({
      success: true,
      message: 'Visiteur créé.',
      visiteur,
      id: visiteur._id,
      estNouveau: true
    });
  } catch (err) {
    console.error('❌ Erreur creerVisiteur :', err.message);
    res.status(500).json({ success: false, message: err.message });
  }
};

const listerVisiteurs = async (req, res) => {
  try {
    const page  = parseInt(req.query.page)  || 1;
    const limit = parseInt(req.query.limit) || 20;
    const skip  = (page - 1) * limit;
    const [total, visiteurs] = await Promise.all([
      Visiteur.countDocuments(),
      Visiteur.find().sort({ createdAt: -1 }).skip(skip).limit(limit),
    ]);
    res.json({ success: true, total, page, pages: Math.ceil(total / limit), visiteurs });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

const getVisiteur = async (req, res) => {
  try {
    const visiteur = await Visiteur.findById(req.params.id);
    if (!visiteur) return res.status(404).json({ success: false, message: 'Visiteur introuvable.' });
    const [visites, documents] = await Promise.all([
      Visite.find({ visiteurId: visiteur._id }).sort({ heureEntree: -1 }),
      Document.find({ visiteurId: visiteur._id }),
    ]);
    res.json({ success: true, visiteur: { ...visiteur.toObject(), visites, documents } });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

const modifierVisiteur = async (req, res) => {
  try {
    const visiteur = await Visiteur.findByIdAndUpdate(
      req.params.id,
      req.body,
      { new: true, runValidators: true }
    );
    if (!visiteur) return res.status(404).json({ success: false, message: 'Visiteur introuvable.' });
    res.json({ success: true, message: 'Visiteur mis à jour.', visiteur });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

// 👇 NOUVELLE FONCTION : suppression en cascade
const supprimerVisiteur = async (req, res) => {
  try {
    const visiteurId = req.params.id;

    const visiteur = await Visiteur.findById(visiteurId);
    if (!visiteur) {
      return res.status(404).json({ success: false, message: 'Visiteur introuvable.' });
    }

    // Supprimer toutes les visites et documents liés
    await Promise.all([
      Visite.deleteMany({ visiteurId }),
      Document.deleteMany({ visiteurId })
    ]);

    // Supprimer le visiteur
    await Visiteur.findByIdAndDelete(visiteurId);

    res.json({
      success: true,
      message: 'Visiteur et toutes ses données associées supprimés avec succès.'
    });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

module.exports = {
  creerVisiteur,
  rechercherParNIN,
  rechercherParTelephone: rechercherParNIN,
  listerVisiteurs,
  getVisiteur,
  modifierVisiteur,
  supprimerVisiteur
};

File: controllers/visiteur_controller.py
from typing import Optional
from datetime import datetime
from fastapi import HTTPException
from pydantic import BaseModel
from bson import ObjectId
from models.visiteur import Visiteur
from models.visite import Visite
from models.document import DocumentScan


class VisiteurBody(BaseModel):
    nom: str
    prenom: str
    date_naissance: Optional[str] = None
    numero_piece: str
    type_piece: str = "CNI"


class VisiteurUpdate(BaseModel):
    nom: Optional[str] = None
    prenom: Optional[str] = None
    date_naissance: Optional[str] = None
    numero_piece: Optional[str] = None
    type_piece: Optional[str] = None


def _visiteur_dict(v: Visiteur) -> dict:
    d = v.dict(by_alias=False)
    d["id"] = str(v.id)
    d.pop("_id", None)
    return d


async def creer_visiteur(body: VisiteurBody):
    existe = await Visiteur.find_one(Visiteur.numero_piece == body.numero_piece)
    if existe:
        return {
            "success":    True,
            "message":    "Visiteur déjà enregistré.",
            "visiteur":   _visiteur_dict(existe),
            "est_nouveau": False,
        }

    date_naissance = None
    if body.date_naissance:
        try:
            date_naissance = datetime.fromisoformat(body.date_naissance)
        except ValueError:
            pass

    visiteur = Visiteur(
        nom=body.nom,
        prenom=body.prenom,
        date_naissance=date_naissance,
        numero_piece=body.numero_piece,
        type_piece=body.type_piece,
    )
    await visiteur.insert()
    return {
        "success":    True,
        "message":    "Visiteur créé.",
        "visiteur":   _visiteur_dict(visiteur),
        "est_nouveau": True,
    }


async def lister_visiteurs(page: int = 1, limit: int = 20):
    skip = (page - 1) * limit
    # ✅ Beanie : find().count() et non Visiteur.count()
    total = await Visiteur.find().count()
    visiteurs = (
        await Visiteur.find()
        .sort(-Visiteur.created_at)
        .skip(skip)
        .limit(limit)
        .to_list()
    )
    return {
        "success":  True,
        "total":    total,
        "page":     page,
        "pages":    -(-total // limit),
        "visiteurs": [_visiteur_dict(v) for v in visiteurs],
    }


async def get_visiteur(visiteur_id: str):
    try:
        oid = ObjectId(visiteur_id)
    except Exception:
        raise HTTPException(status_code=400, detail="ID invalide.")

    visiteur = await Visiteur.get(oid)
    if not visiteur:
        raise HTTPException(status_code=404, detail="Visiteur introuvable.")

    visites   = await Visite.find(Visite.visiteur_id == oid).sort(-Visite.heure_entree).to_list()
    documents = await DocumentScan.find(DocumentScan.visiteur_id == oid).to_list()

    d = _visiteur_dict(visiteur)
    d["visites"]   = [v.dict(by_alias=False) for v in visites]
    d["documents"] = [doc.dict(by_alias=False) for doc in documents]
    return {"success": True, "visiteur": d}


async def modifier_visiteur(visiteur_id: str, body: VisiteurUpdate):
    try:
        oid = ObjectId(visiteur_id)
    except Exception:
        raise HTTPException(status_code=400, detail="ID invalide.")

    visiteur = await Visiteur.get(oid)
    if not visiteur:
        raise HTTPException(status_code=404, detail="Visiteur introuvable.")

    update_data = {k: v for k, v in body.dict().items() if v is not None}
    for key, val in update_data.items():
        setattr(visiteur, key, val)

    visiteur.updated_at = datetime.utcnow()
    await visiteur.save()
    return {"success": True, "message": "Visiteur mis à jour.", "visiteur": _visiteur_dict(visiteur)}